app-root, lib-user: Use url search param for stats date range and project state

mcbouslog opened the pull request:
## Package
- app-root
- lib-user

## Describe your changes
- app-root
  - remove stats date range context
  - remove stats project context
  - add search param for stats start date and end date state to UserStats, Certificate, and GroupStats
  - add search param for stats project state to UserStats, Certificate, and GroupStats
- lib-user
  - refactor UserStats, Certificate, and GroupStats with new dateRange object (includes start date and end date)
  - refactor MainContent for new dateRange object and new Select onChange functions
  - refactor `getDateInterval` utility
  - add `getDateRangeSelectOptions` utility
  - set 'All Time' start date as stats source (user or group) `created_at` date
  - refactor BarChart for new dateRange object
  - refactor `getDateRangeLabel` helper
  - update stories and tests as needed
  - update dev app for stats state per url search param

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX?
- What user actions should my reviewer step through to review this PR?
  - go to your user stats page
    - change the project
    - change the date range
    - generate a certificate
  - go to a group stats page
    - change the project
    - change the date range
    - go to the full contributors page (no related changes, but should work as expected with stats from all time and all projects)
- Which storybook stories should be reviewed? MainContent, BarChart
- Are there plans for follow up PR’s to further fix this bug or develop this feature? yes, see linked GitHub project

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

## Refactoring
- [ ] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

eatyourgreens commented on packages/app-root/src/app/groups/[groupId]/GroupStatsContainer.js:
```diff
@@ -1,16 +1,77 @@
+  // set start date per query params or default to 7 days ago
+  let selectedStartDate = startDate
+  if (!selectedStartDate) {
+    selectedStartDate = new Date(new Date().setDate(new Date().getDate() - 6)).toISOString().substring(0, 10)
```
This might be more readable, and faster since it only uses one new Date object.
```suggestion
    const today = newDate()
    const 7DaysAgo = today.setUTCDate(today.getUTCDate() - 7)
    selectedStartDate = 7DaysAgo.toISOString().substring(0, 10)
```

mcbouslog replied:
Yes! I forgot to include the getUTCDate here. I think I've included it in other spots (like BarChart's [`getCompleteData`](packages/lib-user/src/components/shared/BarChart/helpers/getCompleteData.js)), but please let me know if you notice any other places I've missed it. Thank you!

eatyourgreens replied:
👍🏻

By the way, `new Date` is redundant here, since `date.setUTCDate()` returns a Date object. 

https://github.com/zooniverse/front-end-monorepo/blob/db295fdd5a076c12f47911fe37401563c6d4f32c/packages/lib-user/src/components/shared/BarChart/helpers/getCompleteData.js#L13

eatyourgreens replied:
Just realised my code suggestion was wrong, so I rewrote it. I always assume, wrongly, that Date methods return the modified Date.

mcbouslog replied:
Just added [1145d22](https://github.com/zooniverse/front-end-monorepo/pull/6160/commits/1145d2244dcab2c28b9495d5c9e666b67b44ae3f) and [9b55f9f](https://github.com/zooniverse/front-end-monorepo/pull/6160/commits/9b55f9fa94347025ac7a6c1818bf0574b762413e) to refactor as noted as well as a few other places.

goplayoutside3 commented on packages/lib-user/src/components/shared/MainContent/MainContent.js:
```diff
@@ -44,6 +42,10 @@ const StyledTab = styled(Button)`
 `
 
 const DEFAULT_HANDLER = () => true
+const DEFAULT_DATE_RANGE = {
+  endDate: '',
+  startDate: ''
+}
```
Question about DEFAULT_DATE_RANGE throughout lib-user. In app-root page.js files, if `searchParams.end_date` and/or `searchParams.start_date` are undefined, the end date is defined as today and the start date if defined as 7 days ago by default. Is there a case where it makes sense for `endDate` or `startDate` to be empty strings in lib-user? If you're needing to check whether either variable exists in `getDateInverval()`, I suggest using `null` instead.

mcbouslog replied:
Sounds good, updated with [1bedbf6](https://github.com/zooniverse/front-end-monorepo/pull/6160/commits/1bedbf6e92ee67c315cc07afa87daef6f3e69681)

goplayoutside3 reviewed (commented):
This is looking really good! I like that you're handling url params in app-root and having lib-user components respond to it. One scenario I noticed though --

Let's say a user clicks on the More Stats button on their homepage, or clicks on a group link from their homepage. When either of those pages load, the default date range is Last 7 Days (as expected), however nothing is added to the url. In other cases when Last 7 Days is selected from the MainContent dropdown by the user, a `start_date` is added to the url. What do you think about automatically adding `start_date` for Last 7 Days upon page load for GroupStatsContainer and UserStatsContainer? I think the date selector and url should always be in sync.

Lastly, I agreed with Jim's comment about timezones. Good to do a double check anywhere you're calculating Last [however many] Days, even in `getDateInverval()`.

goplayoutside3 commented:
🤔 With the latest changes, I'm getting an error `.setUTCDate(...).toISOString is not a function` on app-root homepage run locally.

eatyourgreens commented:
>🤔 With the latest changes, I'm getting an error .setUTCDate(...).toISOString is not a function on app-root homepage run locally.

That's the same error I made: `date.setUTCDate()` doesn't return a Date, so it can't be chained. You have to do something like:

```js
date.setUTCDate(newDate);
formattedDate = date.toISOString().slice(0,10);
```

mcbouslog commented:
> 🤔 With the latest changes, I'm getting an error `.setUTCDate(...).toISOString is not a function` on app-root homepage run locally.

Should be ok now, let me know if not.

mcbouslog commented on packages/lib-user/src/components/Certificate/Certificate.stories.js:
```diff
@@ -23,31 +26,40 @@ export const ProjectSpecific = {
     displayName: 'Example User',
     hours: 45,
     login: 'testUser',
```
Nice catch! Updated with [4158432](https://github.com/zooniverse/front-end-monorepo/pull/6160/commits/4158432fa0dc95a0b7a0b56a6185dc578d354e42)

mcbouslog commented:
- moved `getDateRangeSelectOptions` to MainContent/helpers
- refactored to update url with start_date search param if missing on user stats page or group stats page load
- refactored UserHome StatsTabContainer `getDateInterval` per date range changes

goplayoutside3 reviewed (approved):
Thanks for the updates! Looks good to merge
